details/encryption.md

```diff
@@ -0,0 +1,382 @@
+helper immediately preceding them.  Thus, the helper to the right of the current
+helper is identified as $P_{i+1}$.  As this is a ring, the helper to the right
+of $P_3$ is not $P_4$, but $P_1$.  Similarly, the previous helper is identified
+as $P_{i-1}$ and the helper to the left of $P_1$ is $P_3$.
+
+In this view, the replicated secret sharing we use has each helper with a share
+in common with the helper to its left ($mk_i$) and a share in common with the
+helper to its right ($mk_{i+1}$).  While it would be equally valid (and
+indistinguishable from this) if we instead used $mk_{i-1}$ and $mk_i$, we use
+$mk_i$ and $mk_{i+1}$ throughout.
+
+</aside>
+
+
+### HPKE Context Creation
+
+The two values destined for each helper are then concatenated and encrypted
+toward that helper using HPKE.  The sender and receiver contexts used by HPKE
+are constructed using an input that captures critical contextual information.
+
+The `info` input for HPKE shall include a distinguishing sequence that is
+composed from:
+
+1. The fixed string "private-attribution", encoded in ASCII,
+
+2. The ASCII serialization of the helper party origin.
+
+3. A single zero-valued byte.
+
+4. The ASCII serialization of the current site origin.
+
+5. A single zero-valued byte.
+
+6. The single-byte key identifier from the key configuration for the helper party.
+
+7. The current epoch, encoded as an two-byte integer in network byte order.
+
+TODO: OHTTP included identifiers for the KEM, KDF, and AEAD in this.  HPKE
+already includes these identifiers in its `suite_id` concept, making that
+somewhat redundant.  Decide whether to add those values here too.
+
+This produces the following process in pseudocode:
+
+```python
+def ipa_info(helper_origin, site_origin, key_id, epoch):
+    return concat(encode_str("private-attribution"),
+              ascii_origin(helper_origin),
+              encode(0, 1),
+              ascii_origin(site_origin),
+              encode(0, 1),
+              encode(key_id, 1),
+              encode(epoch, 2))
+```
+
+### Encryption
+
+The encryption process is applied by the user agent by creating a sender context
+that uses the helper public key, the `info` string, and - implicitly - a choice
+of KDF and AEAD from the helper configuration.  This context is used to seal a
+single message that is the simple concatenation of the two shares for that
+helper and empty associated data.  This is concatenated with the encapsulation
+key to produce the message for that helper.
+
+This process is applied for each helper, as follows:
+
+```python
+info = ipa_info(helper_origin, site_origin, key_id, epoch)
+enc, sctxt = SetupBaseS(pkH, info)
+ct = sctxt.Seal("", concat(mk[i], mk[i+1]))
+enc_mk = concat(enc, ct)
+```
+
+If match keys are not a whole multiple of 4 bits in length, the combined match
+keys will need to be padded to a whole multiple of 8 bits.
+
+The overall process produces an output that consists of an epoch and three
```

Comment:
This is a good point.  I spent a bit of time thinking about this and decided that we would need to deal with that in any interface document.  I don't think that we need a byte-for-byte `info` value in the API...provided that the information needed is exposed.  The site won't be decrypting this data, so they don't need that.  But they will need to pass the information along the pipeline.

If you look at what is in here, we have:

1. The helper party origin, which the site commits to and so there should be no ambiguity.  I've chosen origin here, which is maybe a bit too narrow, so we might want to change how that is identified.  Either way, I don't see this as critical information, except to ensure that helper A doesn't get data that was meant for helper B.  I'm inclined to include it explicitly though, just to avoid ambiguity.  More so if the site identifies a helper party *network* rather than each individual helper.
2. Site origin.  The site should already know who it is.  If not, `window.location.origin` is available.  No harm in providing this though.
3. Key identifier.  This really is something that will need to be carried.  The key configuration that the browser chooses might differ from what other browsers choose.
4. Epoch.  Though this could be derived from `new Date()`, I think we should include it explicitly.  This value has to be the same for all three encrypted match keys.

Providing this in a single blob will be inefficient and awkward.  Sites won't want to log all of this, and we shouldn't force them to parse out what they need to log.  And the data won't be provided to helpers as a single blob like that.

I think we should be providing the information as a struct, something like:

```json
{ "epoch": 5,
  "helper1.example": { "encryptedMatchKey": ...arraybufferview..., keyId: 25 },
  "helper2.example.net": { "encryptedMatchKey": ...arraybufferview..., keyId: 1 },
  "helper3.example.com": { "encryptedMatchKey": ...arraybufferview..., keyId: 231 } }
```